runtime/runtime/src/bandwidth_scheduler/scheduler.rs

```diff
@@ -0,0 +1,737 @@
+    /// Was this a chunk missing in the current block?
+    /// If the last chunk was missing, receipts sent previously were not processed.
+    pub last_chunk_missing: bool,
+    /// Is this the chosen shard which is allowed to send receipts to other fully congested shards?
+    pub is_allowed_sender_shard: bool,
+}
+
+/// A representation of `BandwidthRequest` used by the bandwidth scheduler.
+/// Instead of absolute values of bandwidth, this struct contains consecutive
+/// increases in bandwidth. The scheduler will go over the increases and either
+/// grant them or refuse them.
+/// The increases already take into account the base bandwidth that is granted on all links.
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SchedulerBandwidthRequest {
+    /// Request to send receipts between those two shards.
+    pub link: ShardLink,
+    /// Requests to increase the granted bandwidth on the link.
+    pub bandwidth_increases: VecDeque<Bandwidth>,
+}
+
+impl SchedulerBandwidthRequest {
+    pub fn new(
+        sender_shard: ShardId,
+        bandwidth_request: &BandwidthRequest,
+        params: &BandwidthSchedulerParams,
+        mapping: &SchedulerShardMapping,
+    ) -> Option<Self> {
+        let Some(sender_index) = mapping.get_index_for_shard_id(sender_shard) else {
+            // Request from a shard that is not in the current set of shards.
+            return None;
+        };
+        let Some(receiver_index) =
+            mapping.get_index_for_shard_id(bandwidth_request.to_shard.into())
+        else {
+            // Request to a shard that is not in the current set of shards.
+            return None;
+        };
+        let link = ShardLink::new(sender_index, receiver_index);
+
+        let mut bandwidth_increases = VecDeque::new();
+
+        // Keeps track of the total bandwidth that would be granted by the requested increases.
+        // Base bandwidth is already granted on all links, so we start with that.
+        let mut current_total = params.base_bandwidth;
+
+        let request_values = BandwidthRequestValues::new(params).values;
+        for bit_idx in 0..bandwidth_request.requested_values_bitmap.len() {
+            if bandwidth_request.requested_values_bitmap.get_bit(bit_idx) {
+                // Request for the total value of bandwidth that should be granted on the link.
+                let requested_value = request_values[bit_idx];
+                if requested_value <= current_total {
+                    continue;
+                }
+                // Convert the absolute value to a bandwidth increase.
+                bandwidth_increases.push_back(requested_value - current_total);
+                current_total = requested_value;
+            }
+        }
+
+        if bandwidth_increases.is_empty() {
+            return None;
+        }
+
+        Some(Self { link, bandwidth_increases })
+    }
+}
+
+/// Iterator over all links from senders to receivers.
+pub struct LinksIterator {
```

Comment:
Is that to avoid nested for loop? Do you need to preserve the state of iteration somewhere?

Reply:
It was mainly done because iterating over `&self.senders` borrowed self and I couldn't call any function that needed `&mut self`, so for example this:
```rust
for sender in self.senders.iter() {
    for receiver in self.receivers.iter() {
        let link = ShardLink::new(*sender, *receiver);
        let _ = self.try_grant_bandwidth(&link, 100);
    }
}
```
wouldn't work because `try_grant_bandwidth` needs a `&mut self`, but the `for` already borrowed `&self` immutably.

But now that `senders` and `receivers` are `Rc<[]>`, looks like this would work:
```
for sender in self.senders.clone().iter() {
    for receiver in self.receivers.clone().iter() {
        let link = ShardLink::new(*sender, *receiver);
        let _ = self.try_grant_bandwidth(&link, 100);
    }
}
```
Maybe that's better than creating own iterator implementation